---
title: "Clena, a tool for differentially expressed genes cluster-pathway enrichment analysis and visualization"
author: "Zhilong Jia, Zhigang Luo and Michael Barnes"
date: "`r Sys.Date()`"
output: 
    rmarkdown::html_vignette:
        fig_caption: yes
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

Clena can be used as gene set enrichment analysis of up or down regulated 
genes as well as genes in clusters resulting from various clustering methods. 
And the gene sets could be Pathway, Gene Ontology, Durg or user-defined gene-sets. 
Accordingly, this tools can be used for pathway enrichment, drug repositioning 
and so on.  

## Input data
Note: all the gene names should be gene SYMBOLs.

- the gene expression profiling of differentially expressed genes. It can be 
matrix with gene in row and sample in column, data.frame or ExpressionSet object.
- the background genes. Generally, these are genes in a microarray after filtering
the non-informative genes. A vector.
- the sample labels indicating the labels, like control and disease, of each 
sample. A vector with sample name.

There is an example data in `clena` package. This dataset, from [GSE20163](http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE20163), is about Parkinson's disease.
we have normalised the expression profling using `rma` method, filtered some non-informative genes using `MetaDE` package and analysed the differentially
expressed genes using `limma` package with the p-value 0.05.

```{r import_data}
library(devtools)
load_all("~/clena")
data(PD)
```

There are three objects in the PD dataset.
```{r PD_data, echo=FALSE}
ls()
```

## clena analysis

1. Make Annotation
    There are a variety of gene sets in the `clena` package, collected from 
    [MSigDB](http://www.broadinstitute.org/gsea/msigdb/index.jsp) and [Enrichr](
    http://amp.pharm.mssm.edu/Enrichr/). They are `r dir(system.file("data", package="clena"), pattern="*gmt")`. with gene-sets, user can get the genes-specific gene
    sets matrix as showed in the following example.
    
```{r gene_list_annotation}
annoGMT <- "c2.cp.kegg.v4.0.symbols.gmt"
annofile <- system.file("data", annoGMT, package="clena")
# the DEG gene-sets matrix
anno <- gene2set(annofile, rownames(DEexprs))
# the background gene gene-sets matrix
annotationGenesPop <- gene2set(annofile, BGgenes)
annotationGenesPop <- annotationGenesPop[,colnames(anno)]
```
    
    Here the gene-sets can be user-defined gene-sets formatted gmt, too.
    
2. Run clena
    Setting some parameters for clena.
```{r clena, results="hide"}
# the number of clusters. It could be a vector.
nClust <- 2:3
# the number of cores. As leveraging the `doMC` package, clena can be used in unix-like system only.
ncore <- 2
# the clustering methods
clMethods <- c("hierarchical","kmeans")
# the distance metric
metric <- "correlation"
# the agglomeration method used for hierarchical clustering (hierarchical and agnes)
method <- "complete"

# the clena analysis
clena_result <- clena(DEexprs, nClust=nClust, clMethods=clMethods, metric=metric,
                      method=method,  annotation=anno, sampleLabel=sampleLabel, 
                      ncore=ncore, annotationGenesPop=annotationGenesPop, verbose=TRUE)
```


## Analysing result of clena
    After completing the clena analysis, user can use `summary` to see the summary 
    of the result of clena.
    

```{r clena_result_analysis}
summary(clena_result)
score <- optCluster(clena_result)
```

```{r score, echo=FALSE, results='asis'}
knitr::kable(score)
```

```{r enrichment}
enrichment.table <- enrichment(clena_result, "kmeans", "3")
```


the enrichment graph is plotted by `heatmapPEI`

```{r heatmapPEI, fig.width=8, fig.height=5, fig.cap="KEGG Pathway enrichment"}
heatmapPEI(clena_result, "kmeans", "3")
```

the heatmap of expression profiling with clusters is plotted by `heatmapCluster`

```{r heatmapCluster, fig.width=9, fig.height=9, fig.cap="Heatmap of expression profiling with clusters"}
heatmapCluster(clena_result, "kmeans", "3")
```

## Bug Report
    github

The END
